core/api/views/cp_records.py

```diff
@@ -122,9 +129,9 @@ def _set_chemical_items_dict(
             )
             if chemical_key not in existing_items_dict:
                 cp_record_data = {
-                    "country_programme_report_id": cp_report.id,
-                    "substance_id": chemical.id if row.substance else None,
-                    "blend_id": chemical.id if row.blend else None,
+                    "country_programme_report": cp_report,
+                    "substance": chemical if row.substance else None,
+                    "blend": chemical if row.blend else None,
```

Comment:
Using the object directly instead of the id doesn't make another query for the whole substance/blend.